WcaOnRails/app/controllers/competitions_controller.rb

```diff
@@ -526,6 +526,20 @@ def embedable_map
     render layout: false
   end
 
+  def bookmark
+    @competition = competition_from_params
+    if !BookmarkedCompetition.exists?(competition: @competition, user: current_user)
+      BookmarkedCompetition.create(competition: @competition, user: current_user)
+    end
+    head :ok
+  end
+
+  def unbookmark
+    @competition = competition_from_params
+    BookmarkedCompetition.where(competition: @competition, user: current_user).destroy_all
```

Comment:
nit: `destory_all` can silently fail to actually delete things (if validations fail). We wrote our own `destory_all!` to use instead to avoid this problem: https://github.com/jfly/worldcubeassociation.org/blob/9e54ed58941020af47b8b0bad6112283b7931ef5/WcaOnRails/config/initializers/monkey_patches.rb#L65-L68

(That said, you could probably use `find(...).destroy!` here instead, since we only expect a user to have 1 bookmark per competition.)

Reply:
> Does destroy_all! work at the end of a query like this?

Darn! I thought it did, but it sounds like it does not =( I think you could do a `.each(&:destroy!)` instead?

> I wanted to guarantee that the competition is not bookmarked after this method is called.

SGTM!

Reply:
Thanks, done.